test/proxy-upgrades/test-upgrade.ts

```diff
@@ -0,0 +1,267 @@
+}
+
+describe("Upgradeability", function () {
+  // TODO
+  let baseImpl: SmartWallet;
+  let walletFactory: WalletFactory;
+  let entryPoint: EntryPoint;
+  let token: MockToken;
+  let multiSend: MultiSend;
+  let storage: StorageSetter;
+  let owner: string;
+  let bob: string;
+  let charlie: string;
+  let userSCW: any;
+  let handler: DefaultCallbackHandler;
+  const VERSION = "1.0.4";
+  const create2FactoryAddress = "0xce0042B868300000d44A59004Da54A005ffdcf9f";
+  let accounts: any;
+
+  /* const domainType = [
+      { name: "name", type: "string" },
+      { name: "version", type: "string" },
+      { name: "verifyingContract", type: "address" },
+      { name: "salt", type: "bytes32" },
+    ]; */
+
+  beforeEach(async () => {
+    accounts = await ethers.getSigners();
+    const addresses = await ethers.provider.listAccounts();
+    const ethersSigner = ethers.provider.getSigner();
+
+    owner = await accounts[0].getAddress();
+    bob = await accounts[1].getAddress();
+    charlie = await accounts[2].getAddress();
+    // const owner = "0x7306aC7A32eb690232De81a9FFB44Bb346026faB";
+
+    const EntryPoint = await ethers.getContractFactory("EntryPoint");
+    entryPoint = await EntryPoint.deploy();
+    await entryPoint.deployed();
+    console.log("Entry point deployed at: ", entryPoint.address);
+
+    const DefaultHandler = await ethers.getContractFactory(
+      "DefaultCallbackHandler"
+    );
+    handler = await DefaultHandler.deploy();
+    await handler.deployed();
+    console.log("Default callback handler deployed at: ", handler.address);
+
+    const BaseImplementation = await ethers.getContractFactory("SmartAccount");
+    baseImpl = await BaseImplementation.deploy(entryPoint.address);
+    await baseImpl.deployed();
+    console.log("base wallet impl deployed at: ", baseImpl.address);
+
+    const WalletFactory = await ethers.getContractFactory(
+      "SmartAccountFactory"
+    );
+    walletFactory = await WalletFactory.deploy(
+      baseImpl.address,
+      handler.address
+    );
+    await walletFactory.deployed();
+    console.log("wallet factory deployed at: ", walletFactory.address);
+
+    const MockToken = await ethers.getContractFactory("MockToken");
+    token = await MockToken.deploy();
+    await token.deployed();
+    console.log("Test token deployed at: ", token.address);
+
+    const Storage = await ethers.getContractFactory("StorageSetter");
+    storage = await Storage.deploy();
+    console.log("storage setter contract deployed at: ", storage.address);
+
+    const MultiSend = await ethers.getContractFactory("MultiSend");
+    multiSend = await MultiSend.deploy();
+    console.log("Multisend helper contract deployed at: ", multiSend.address);
+
+    console.log("mint tokens to owner address..");
+    await token.mint(owner, ethers.utils.parseEther("1000000"));
+  });
+
+  // describe("Wallet initialization", function () {
+  it("Should set the correct states on proxy", async function () {
```

Comment:
Could add a test where we upgrade while decrementing the _gap and introduce some new state and verify everything still behaves expected.

Reply:
there is test case for this. state-upgrade-base.ts